core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1689,6 +1690,123 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
+      List<List<RexNode>> disjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::disjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(disjunctions.get(0));
+      disjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return and(simplifiedChildren);
+      } else {
+        // Get the predicates that without the common part.
+        disjunctions.forEach(p -> p.removeAll(common));
+        if (disjunctions.stream().anyMatch(List::isEmpty)) {
+          // (a || b) && (a || b || c || d) => (a || b)
+          return or(common);
+        } else {
+          // (a || b || c || d) && (a || b || e || f) =>
+          // (a || b) || ((c || d) && (e || f))
+          List<RexNode> splice = disjunctions.stream()
+              .map(this::or)
+              .collect(Collectors.toList());
+          return or(or(common), and(splice));
+        }
+      }
+    }
+    case OR: {
+      // Split the children to get the conjunctions predicates.
+      List<List<RexNode>> conjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::conjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(conjunctions.get(0));
+      conjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return or(simplifiedChildren);
+      } else {
+        // Get the predicates that without the common part.
+        conjunctions.forEach(p -> p.removeAll(common));
+        if (conjunctions.stream().anyMatch(List::isEmpty)) {
+          // (a && b) || (a && b && c) => (a && b)
+          return and(common);
+        } else {
+          // (a && b && c && d) || (a && b && e && f) =>
+          // (a && b) && ((c && d) || (e && f))
+          List<RexNode> splice = conjunctions.stream()
+              .map(this::and)
+              .collect(Collectors.toList());
+          return and(and(common), or(splice));
+        }
+      }
+    }
+    default:
+      return call;
+    }
+  }
+
+  private RexNode and(RexNode... nodes) {
+    return and(ImmutableList.copyOf(nodes));
+  }
+
+  private RexNode and(List<RexNode> nodes) {
+    if (nodes.size() == 1) {
+      return nodes.get(0);
+    } else {
+      return rexBuilder.makeCall(SqlStdOperatorTable.AND,
+          ImmutableList.copyOf(nodes));
+    }
+  }
+
+  private RexNode or(RexNode... nodes) {
+    return or(ImmutableList.copyOf(nodes));
+  }
+
+  private RexNode or(List<RexNode> nodes) {
+    if (nodes.size() == 1) {
+      return nodes.get(0);
+    } else {
+      return rexBuilder.makeCall(SqlStdOperatorTable.OR,
+          ImmutableList.copyOf(nodes));
+    }
+  }
```

Comment:
Do these `and` and `or` belong to `RexSimplify`?
I believe we have the same thing in `RexUtil`

Reply:
Sure, I'll take a look.

Reply:
Maybe RexUtil.composeDisjunction / RexUtil.composeConjunction

Reply:
Sure~ revised.